esp-hal/src/i2s/master/camera.rs

```diff
@@ -0,0 +1,404 @@
+            // Enable slave receiver mode
+            w.rx_slave_mod()
+                .set_bit()
+                // Receive left-channel data first.
+                .rx_right_first()
+                .clear_bit()
+                // Place left-channel data at the MSB in the FIFO
+                .rx_msb_right()
+                .clear_bit()
+                // Do not enable receiver in Philips standard mode
+                .rx_msb_shift()
+                .clear_bit()
+                // Do not enable receiver’s mono mode in PCM standard mode
+                .rx_mono()
+                .clear_bit()
+                // Do not enable receiver in PCM standard mode
+                .rx_short_sync()
+                .clear_bit()
+        });
+
+        // ADC/LCD/camera configuration register
+        regs.conf2().modify(|_, w| {
+            // Enable LCD mode.
+            w.lcd_en()
+                .set_bit()
+                // Enable camera mode.
+                .camera_en()
+                .set_bit()
+        });
+
+        // Configure clock divider
+        regs.clkm_conf().modify(|_, w| unsafe {
+            w.clkm_div_a()
+                .bits(0) // Fractional clock divider’s denominator value (6 bits)
+                .clkm_div_b()
+                .bits(0) // Fractional clock divider’s numerator value. (6 bits)
+                .clkm_div_num()
+                .bits(2) // I2S clock divider’s integral value. (8 bits)
+        });
+
+        regs.fifo_conf().modify(|_, w| unsafe {
+            // Enable I2S DMA mode.
+            w.dscr_en()
+                .set_bit()
+                // Receive FIFO mode configuration bit. (3 bits)
+                .rx_fifo_mod()
+                .bits(data_format as _)
+                // The bit should always be set to 1.
+                .rx_fifo_mod_force_en()
+                .set_bit()
+        });
+
+        regs.conf_chan().modify(|_, w| unsafe {
+            // I2S receiver channel mode configuration bit. (2 bits)
+            w.rx_chan_mod().bits(1)
+        });
+
+        // Configure the bit length of I2S receiver channel. (6 bits)
+        regs.sample_rate_conf()
+            .modify(|_, w| unsafe { w.rx_bits_mod().bits(16) });
+
+        // Synchronize signals into the receiver in double sync method.
+        regs.timing().write(|w| w.rx_dsync_sw().set_bit());
+
+        // Default these signals to high in case user doesn't provide them.
+        I2S::v_sync_signal().connect_to(Level::High);
+        I2S::h_sync_signal().connect_to(Level::High);
+        I2S::h_enable_signal().connect_to(Level::High);
+
+        let rx_channel = ChannelRx::new(channel.map(|ch| ch.degrade()));
+
+        Self {
+            _i2s: i2s,
+            rx_channel,
+        }
+    }
+}
+
+impl<I2S> Camera<'_, I2S>
+where
+    I2S: RegisterAccess + ExtendedSignals,
+{
+    /// Configures the data pins for the camera interface.
+    #[allow(clippy::too_many_arguments)]
+    pub fn with_data_pins<
```

Comment:
I know it's a bit tedious but these should be split out into a function per pin, per [guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/DEVELOPER-GUIDELINES.md).

Reply:
doesn't that then run the (big) risk of a user accidentally missing a pin? we can't catch such an oversight and there's (probably?) never any valid reason to use this functionality with anything but the full set of data pins?
also, the current API is aligned with `lcd_cam::cam`, making it easier to move between the two. shouldn't that be valued higher (and both be refactored together at a later step by maintainers when they decide on what the APIs should look like for stabilisation)?

Reply:
> the current API is aligned with `lcd_cam::cam`

https://github.com/esp-rs/esp-hal/pull/3237 :smile:

Reply:
as discussed [on matrix](https://matrix.to/#/!LdaNPfUfvefOLewEIM:matrix.org/$cUiQVpZ8Qi45aeLV5ch0WzeUAx4X-InBZps_TUGx6t0?via=matrix.org&via=tchncs.de&via=mozilla.org) i've now added individual `with_d[0..7]` functions in addition to `with_data_pins`